py3status/modules/speedtest.py

```diff
@@ -0,0 +1,299 @@
+
+STRING_NOT_INSTALLED = "not installed"
+MISSING_PLACEHOLDER = "{download} or {upload} placeholder required"
+
+
+class Py3status:
+    """
+    """
+
+    # available configuration parameters
+    button_refresh = 2
+    button_share = None
+    format = (
+        u"\u25cf [{ping} ms] [↑ {upload} {upload_unit}] [↓ {download} {download_unit}]"
+    )
+    server_id = None
+    si_units = False
+    sleep_timeout = 5
+    thresholds = {
+        "download": [(0, "bad"), (1024, "degraded"), (1024 * 1024, "good")],
+        "ping": [(200, "bad"), (150, "orange"), (100, "degraded"), (10, "good")],
+        "quality": [
+            ("ok", "good"),
+            ("bad", "degraded"),
+            ("faster", "good"),
+            ("slower", "degraded"),
+            ("faster", "good"),
+        ],
+        "upload": [(0, "bad"), (1024, "degraded"), (1024 * 1024, "good")],
+    }
+    timeout = 10
+    unit_bitrate = "MB/s"
+    unit_size = "MB"
+
+    def post_config_hook(self):
+        if not self.py3.check_commands("speedtest-cli"):
+            raise Exception(STRING_NOT_INSTALLED)
+
+        self.first_run = True
+
+        self.placeholders = self.py3.get_placeholders_list(self.format)
+
+        # if download* or upload* missing, run complete test
+        no_upload = ""
+        no_download = ""
+        if any(
+            key.startswith(x)
+            for x in ["download", "upload"]
+            for key in self.placeholders
+        ):
+            if not any(key.startswith("upload") for key in self.placeholders):
+                no_upload = "--no-upload"
+
+            if not any(key.startswith("download") for key in self.placeholders):
+                no_download = " --no-download"
+
+        server = ""  # if not specified, nearest server is use
+        if self.server_id and int(self.server_id):
+            server = "--server {}".format(self.server_id)
+
+        share = ""
+        if self.button_share and all(
+            x in self.placeholders for x in ["download", "upload"]
+        ):
+            share = "--share"
+
+        self.command = "speedtest-cli --json --secure --timeout {} {} {} {} {}".format(
+            self.timeout, no_upload, no_download, server, share
+        )
+
+    def _is_running(self):
+        try:
+            self.py3.command_output(["pgrep", "speedtest-cli"])
+            return True
+        except self.py3.CommandError:
+            return False
+
+    def _get_speedtest_data(self):
+        return loads(self.py3.command_output(self.command)) or None
+
+    def speedtest(self):
+        speedtest_data = {}
+        self.url = None
+        if self.first_run:
+            self.first_run = False
+            cached_until = self.py3.CACHE_FOREVER
+        else:
+            current_data = {}
+            previous_data = {}
+            cached_until = self.py3.time_in(self.sleep_timeout)
+
+            if not self._is_running():
+                previous_data = self.py3.storage_get("speedtest_data")
```

Comment:
I assume you want `or` at end with something else to reduce additional checkings below.